### PR TITLE
Add minimum score criteria for AI search results

### PR DESCRIPTION
**Pull request by sogue**
## Purpose

A filtering mechanism based on minimum score criteria for documents. By setting a minimum value for both `search_score` and `reranker_score`. This aims to improve the relevance of documents sent to the AI, that only the most relevant information is sent for further processing.

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[ ] Yes
[x] No
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[X] Yes
[ ] No
```

## Type of change

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](../CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [x] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [ ] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.
De

### CODE REVIEW

**Comment by sogue** [comment]
@zedhaque No, enabling semantic ranker and semantic captions isn't required for this change to work. The filtering based on search_score operates independently of those features. But for reranker_score it is needed. If one of the scores don't meet the threshold, documents won't be sent to the AI for further processing. Instead, you could either retrieve more documents hoping some meet the criteria or inform the user that no relevant information was found, depending on your implementation.

File: app/frontend/src/pages/ask/Ask.tsx [hunk 1]
```diff
@@ -259,6 +279,30 @@ export function Component(): JSX.Element {
...
+                    className={styles.chatSettingsSeparator}
+                    label="Min Search Score"
+                    min={0}
+                    max={1}
+                    step={0.1}
+                    defaultValue={minSearchScore}
+                    onChange={onMinSearchScoreChange}
+                    showValue
+                    snapToStep
+                />
```
Comment (pamelafox) [inline app/frontend/src/pages/ask/Ask.tsx:292]: I wonder if 0.1/0.2/0.3 is detailed enough, given it only goes from 0 to 1? More importantly, as far as I understand, this score varies based on whether the mode is vector, hybrid, or text:
https://learn.microsoft.com/en-us/azure/search/hybrid-search-ranking#scores-in-a-hybrid-search-results
So I think it'd have to be an `input type=number` or whatever the equivalent FluentUI component is.

Please correct me if I'm wrong though.

File: app/frontend/src/pages/ask/Ask.tsx [hunk 2]
```diff
@@ -259,6 +279,30 @@ export function Component(): JSX.Element {
                     snapToStep
                 />
 
+                <Slider
+                    className={styles.chatSettingsSeparator}
+                    label="Min Search Score"
```
Comment (pamelafox) [inline app/frontend/src/pages/ask/Ask.tsx:284]:
```suggestion
                    label="Minimum search score"
```

File: app/frontend/src/pages/ask/Ask.tsx [hunk 3]
```diff
@@ -259,6 +279,30 @@ export function Component(): JSX.Element {
...
+                    step={0.1}
+                    defaultValue={minSearchScore}
+                    onChange={onMinSearchScoreChange}
+                    showValue
+                    snapToStep
+                />
+
+                <Slider
+                    className={styles.chatSettingsSeparator}
+                    label="Min Reranker Score"
```
Comment (pamelafox) [inline app/frontend/src/pages/ask/Ask.tsx:296]:
```suggestion
                    label="Minimum reranker score"
```

...if we have room for "minimum", I think we probably do?

**Review by pamelafox — CHANGES_REQUESTED** [review]
Thanks so much for the PR! We very much need this functionality. I have a comment about the score range, and there are also failing mypy checks, so please address those. Also will assign a reviewer from search team since this is a search specific change.

**Comment by pamelafox** [comment]
This PR *would* require an update to the learn.microsoft.com tutorial by the way, as that tutorial includes a screenshot of developer settings and an explanation of each setting. FYI @diberry 
I will wait for this to be approved before sending a PR for the tutorial.

**Comment by sogue** [comment]
> Thanks so much for the PR! We very much need this functionality. I have a comment about the score range, and there are also failing mypy checks, so please address those. Also will assign a reviewer from search team since this is a search specific change.

Thanks for the feedback! I've fixed the mypy issues and made the score range input flexible as suggested. The PR is updated and ready for the next review.

File: app/frontend/src/pages/ask/Ask.tsx (hunk 1 above)
Comment (sogue) [reply app/frontend/src/pages/ask/Ask.tsx:280]: Yes, you're right. The score range varies significantly depending on the search method. I've updated the UI to replace the slider with a numeric input field.

**Comment by sogue** [comment]
> Thanks so much for the PR! We very much need this functionality. I have a comment about the score range, and there are also failing mypy checks, so please address those. Also will assign a reviewer from search team since this is a search specific change.

Thanks for the feedback! I've fixed the mypy issues and made the score range input flexible as suggested. The PR is updated and ready for the next review.

File: app/backend/approaches/approach.py [hunk 4]
```diff
@@ -161,7 +163,16 @@ async def search(
                         reranker_score=document.get("@search.reranker_score"),
                     )
                 )
-        return documents
+
+            qualified_documents = [
```
Comment (pamelafox) [inline app/backend/approaches/approach.py:167]: Could you add a unit test in our pytest test suite that checks this logic? You can test the Approach.search() function directly and use a mock of SearchClient.search, we have that mocked in several places in our tests. Let me know if you need further guidance. Thanks!

**Review by pamelafox — COMMENTED** [review]
Thank you, those changes look good. I've asked @mattgotteiner to review.

**Review by mattgotteiner — APPROVED** [review]
Thanks!

**Review by pamelafox — APPROVED** [review]

**Pull request merged**